Check images for true alpha transparency

siliconforks opened the pull request:
Because ImageMagick cannot save images with true alpha transparency in indexed PNG format, we need to check whether an image has true alpha transparency before telling ImageMagick to use the png8 (indexed) format.

Trac ticket: https://core.trac.wordpress.org/ticket/63448

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

siliconforks committed:
Check images for true alpha transparency

Because ImageMagick cannot save images with true alpha transparency in
indexed PNG format, we need to check whether an image has true alpha
transparency before telling ImageMagick to use the png8 (indexed)
format.

See https://core.trac.wordpress.org/ticket/63448

SirLouen commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -480,29 +513,33 @@ protected function thumbnail_image( $dst_w, $dst_h, $filter_name = 'FILTER_TRIAN
+					 * We need to avoid forcing indexed format for images with true alpha transparency,
+					 * because ImageMagick does not support saving an image with true alpha transparency as an indexed PNG.
+					 */
+					if ( Imagick::COLORSPACE_GRAY === $this->image->getImageColorspace() && ! $is_indexed_png_with_true_alpha_transparency ) {
```
I would remove the colorspace_gray check and move this to L531 as the first check.

Like this:

```php
/*
* We need to avoid forcing indexed format for images with true alpha transparency,
* because ImageMagick does not support saving an image with true alpha transparency as an indexed PNG.
*/
if ( ! $is_indexed_png_with_true_alpha_transparency ) {
    $this->image->setOption( 'png:format', 'png8' );
} else {
    $this->image->quantizeImage( 256, $this->image->getColorspace(), 0, false, false );
}
```

siliconforks replied:
Well, if we did that, I think the code would probably work OK - but then there would be no indication of what the purpose of that branch actually is.

As far as I know, the reason that branch is there is specifically to work around an edge case with grayscale images.  Otherwise I think there's no reason to have that branch at all?  (I believe just calling `quantizeImage` would work for non-grayscale images.)  So I would like to retain some indication of this, either in the code itself (so that `setOption` would be called only for grayscale images), or at least mention in a comment the reason for having that branch.

SirLouen replied:
Greyscale condition was only a supposition that `setOption` was the way to deal with them.

That was originally expected like that. But completely out of chance I found that the real right way to index any (both colour and greyscale) 1-bit alpha depth images was using the `setOption`

Quantization is just a way to play around with images with a true alpha transparency. I would simply comment this.

I'm not 100% confident yet, but given that the results in size are identical in 99% of the cases, I would say that `setOption` is some sort of quantization. As I said in my original post, the problem for me was that I could not see when `png8` was viable because I could not identify the true alpha transparency. But now that there is a solution, quantization should be done as the last resort.

PS: I would add tests with all the alpha images we played with, checking if they preserve the alpha bits (theoretically with the current code they should fail because `png8` is trimming them. With the mod I'm suggesting here, I've run the current 9 assertions and they are obviously working correctly as expected.

adamsilverstein replied:
is this feedback resolved already?

siliconforks replied:
I tried eliminating the grayscale check but found that images with a lot of colors might experience significant quality degradation when saving in `png8` format:

https://core.trac.wordpress.org/ticket/63448#comment:86

siliconforks committed:
Remove blank line to make phpcs happy

siliconforks committed:
Test that resizing preserves alpha transparency

siliconforks committed:
Add more comments

The additional comments are based on a suggestion from @SirLouen:

https://core.trac.wordpress.org/ticket/63448#comment:89

(But I have added a bit more detail.)

adamsilverstein commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -444,6 +444,39 @@ protected function thumbnail_image( $dst_w, $dst_h, $filter_name = 'FILTER_TRIAN
+			$is_png                                      = false;
+			$is_indexed_png                              = false;
+			$is_indexed_png_with_alpha_channel           = false;
+			$is_indexed_png_with_true_alpha_transparency = false;
```
can we add a comment here indicating the meaning of true alpha channel (vs alpha channel)

siliconforks replied:
I've added a comment which explains the purpose of all these variables (and mentions what "true alpha" means here).  Does that make things clearer?

adamsilverstein replied:
yes, thanks. I will update the language slightly.

adamsilverstein commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -444,6 +444,39 @@ protected function thumbnail_image( $dst_w, $dst_h, $filter_name = 'FILTER_TRIAN
+							&& defined( 'Imagick::CHANNEL_ALPHA' )
+						) {
+							$alpha_channel_depth = $this->image->getImageChannelDepth( Imagick::CHANNEL_ALPHA );
+							if ( $alpha_channel_depth > 1 ) {
```
this conditional can be included in the if block above

adamsilverstein commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -444,6 +444,39 @@ protected function thumbnail_image( $dst_w, $dst_h, $filter_name = 'FILTER_TRIAN
+
+						if (
+							is_callable( array( $this->image, 'getImageChannelDepth' ) )
+							&& defined( 'Imagick::CHANNEL_ALPHA' )
```
```suggestion
							&& defined( 'Imagick::CHANNEL_ALPHA' )
							&& 1 < $this->image->getImageChannelDepth( Imagick::CHANNEL_ALPHA )
```
something like this.

adamsilverstein commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -444,6 +444,39 @@ protected function thumbnail_image( $dst_w, $dst_h, $filter_name = 'FILTER_TRIAN
 		}
 
 		try {
+			$is_png                                      = false;
```
do we really need all these extra variables? as far as I can tell these are only used once, which makes adding them excessive. can we just remove this part entirely, it seems unrelated.

siliconforks replied:
The reason for all the additional variables:

1. I don't think it is possible to eliminate the `$is_indexed_png_with_true_alpha_transparency` variable because we need to perform the check *before* resizing the image, but we need to use that information *after* the resize.  So I don't see how we can avoid this one.

2. The other variables are (I think) not strictly necessary, but if we eliminate them we will end up duplicating all the checks before and after resizing the image.  Some of these checks are pretty complicated (e.g., `is_callable( array( $this->image, 'getImageProperty' ) ) && '3' === $this->image->getImageProperty( 'png:IHDR.color-type-orig' )`) so it seems simpler just to store the results in variables.

adamsilverstein replied:
makes sense - thanks for clarifying.

siliconforks committed:
Eliminate unneeded $alpha_channel_depth variable

siliconforks committed:
Add comment for variables

adamsilverstein reviewed (approved):
Looks good!

adamsilverstein committed:
Merge branch 'trunk' into 63448-check-for-true-alpha-transparency
